README.md
# nRF BLINKY

nRF Blinky is an example app targeted towards newcomer BLE developers.
This application will demonstrate how to control a LED on an nRF development kit and receive notifications when the button on the kit is pressed and released.

## Nordic LED and Button Service
###### Service UUID: `00001523-1212-EFDE-1523-785FEABCD123`
A simplified proprietary service by Nordic Semiconductor, containing two characteristics one to control LED3 and Button1:
- First characteristic controls the LED state (On/Off).
  - UUID: **`00001525-1212-EFDE-1523-785FEABCD123`**
  - Value: **`1`** => LED On
  - Value: **`0`** => LED Off
- Second characteristic notifies central of the button state on change (Pressed/Released).
  - UUID: **`00001524-1212-EFDE-1523-785FEABCD123`**
  - Value: **`1`** => Button Pressed
  - Value: **`0`** => Button Released

### Requirements:
- An iOS device with BLE capabilities
- A [nRF52](https://www.nordicsemi.com/eng/Products/Bluetooth-low-energy/nRF52-DK) or [nRF51](https://www.nordicsemi.com/eng/Products/nRF51-DK) Dev Kit
- The Blinky example firmware to flash on the Development Kit, there are a few places to acquire that:
- For your conveninence, we have bundled two firmwares in this project under the Firmwares directory.
- To get the latest firmwares and check the source code, you may go directly to our [Developers website](http://developer.nordicsemi.com/nRF5_SDK/) and download the SDK version you need, then you can find the source code and hex files to the blinky demo in the directory `/examples/ble_peripheral/ble_app_blinky/`
-  More information about the nRFBlinky example firmware can be found in the [Info center](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.sdk5.v14.1.0%2Fble_sdk_app_blinky.html)

### Installation and usage:
- Prepare your Development kit.
  - Plug in the Development Kit to your computer via USB.
  - Power On the Development Kit.
  - The Development Kit will now appear as a Mass storage device.
  - Drag (or copy/paste) the appropriate HEX file onto that new device.
  - The Development Kit lEDS will flash and it will disconnect and reconnect.
  - The Development Kit is now ready and flashed with the nRFBlinky example firmware.

- Start Xcode and run build the project against your target iOS Device (**Note:** BLE is not available in the iOS simulator, so the iOS device is a requirement to test the application).
  - Launch the nRFBlinky app on your iOS device.
  - The app will start scanning for nearby peripherals.
  - Select the nRF_Blinky peripheral that appears on screen (**Note:** if the peripheral does not show up, ensure that it's powered on and functional).
  - Your iOS device will now connect to the peripheral and state is displayed on the screen.
  - Changing the value of the Toggle switch will turn LED 3 on or off.
  - Pressing Button 1 on the Development Kit will show the button state as Pressed on the app.
  - Releasing Button 1 will show the state as Released on the App.


CHANGELOG
- **1.1.0**:
    - Improvement: New Nordic look & feel
    - Improvement: Localisatino support with Spanish, Italian, Norwegian, French, Russian, Ukrainian, Romanian, Polish, German and Korean
    - Bugfix: FIxed hint positioning after orientation change
    
- **1.0.2**:
    - Improvement: Blinky peripheral name is now used as the title of the view instead of the default name
    - Improvement: Empty peripherals view resizes and animates better with orientation change
    - Improvement: Fixed a minor detail in the (Wireless by Nordic) text at the bottom of the view to avoid situations where it disappears or does not animate properly with orientation changes

- **1.0.1**:
    - Bugfix: Fixed an issue where swiping back from the peripheral view without actually dismissing would cause the app not to reconnect to the Blinky peripheral.

- **1.0**:
    - Initial implementation


nRFBlinky/Models/BlinkyPeripheral.swift
//
//  BlinkyPeripheral.swift
//  nRFBlinky
//
//  Created by Mostafa Berg on 28/11/2017.
//  Copyright © 2017 Nordic Semiconductor ASA. All rights reserved.
//

import UIKit
import CoreBluetooth

class BlinkyPeripheral: NSObject, CBPeripheralDelegate {
    
    // MARK: - Blinky services and charcteristics Identifiers
    
    public static let nordicBlinkyServiceUUID  = CBUUID.init(string: "00001523-1212-EFDE-1523-785FEABCD123")
    public static let buttonCharacteristicUUID = CBUUID.init(string: "00001524-1212-EFDE-1523-785FEABCD123")
    public static let ledCharacteristicUUID    = CBUUID.init(string: "00001525-1212-EFDE-1523-785FEABCD123")
    
    // MARK: - Properties
    
    public private(set) var basePeripheral    : CBPeripheral
    public private(set) var advertisedName    : String?
    public private(set) var RSSI              : NSNumber
    public private(set) var advertisedServices: [CBUUID]?
    
    // MARK: - Callback handlers
    
    private var ledCallbackHandler: ((Bool) -> (Void))?
    private var buttonPressHandler: ((Bool) -> (Void))?

    // MARK: - Services and Characteristic properties
    
    private var blinkyService       : CBService?
    private var buttonCharacteristic: CBCharacteristic?
    private var ledCharacteristic   : CBCharacteristic?

    init(withPeripheral peripheral: CBPeripheral, advertisementData advertisementDictionary: [String : Any], andRSSI currentRSSI: NSNumber) {
        basePeripheral = peripheral
        RSSI = currentRSSI
        super.init()
        (advertisedName, advertisedServices) = parseAdvertisementData(advertisementDictionary)
        basePeripheral.delegate = self
    }
    
    public func setLEDCallback(_ handler: @escaping (Bool) -> (Void)) {
        ledCallbackHandler = handler
    }

    public func setButtonCallback(_ handler: @escaping (Bool) -> (Void)) {
        buttonPressHandler = handler
    }
    
    public func removeButtonCallback() {
        buttonPressHandler = nil
    }
    
    public func removeLEDCallback() {
        ledCallbackHandler = nil
    }

    public func discoverBlinkyServices() {
        print("Discovering LED Button service...")
        basePeripheral.delegate = self
        basePeripheral.discoverServices([BlinkyPeripheral.nordicBlinkyServiceUUID])
    }
    
    public func discoverCharacteristicsForBlinkyService(_ service: CBService) {
        print("Discovering LED and Button characteristrics...")
        basePeripheral.discoverCharacteristics([BlinkyPeripheral.buttonCharacteristicUUID,
                                            BlinkyPeripheral.ledCharacteristicUUID],
                                           for: service)
    }
    
    public func enableButtonNotifications(_ buttonCharacteristic: CBCharacteristic) {
        print("Enabling notifications for Button characteristic...")
        basePeripheral.setNotifyValue(true, for: buttonCharacteristic)
    }
    
    public func readLEDValue() {
        if let ledCharacteristic = ledCharacteristic {
            print("Reading LED characteristic...")
            basePeripheral.readValue(for: ledCharacteristic)
        }
    }
    
    public func readButtonValue() {
        if let buttonCharacteristic = buttonCharacteristic {
            print("Reading Button characteristic...")
            basePeripheral.readValue(for: buttonCharacteristic)
        }
    }

    public func didWriteValueToLED(_ value: Data) {
        print("LED value written \(value[0])")
        if value[0] == 1 {
            ledCallbackHandler?(true)
        } else {
            ledCallbackHandler?(false)
        }
    }
    
    public func didReceiveButtonNotificationWithValue(_ value: Data) {
        print("Button value changed to: \(value[0])")
        if value[0] == 1 {
            buttonPressHandler?(true)
        } else {
            buttonPressHandler?(false)
        }
    }
    
    public func turnOnLED() {
        writeLEDCharcateristicValue(Data([0x1]))
    }
    
    public func turnOffLED() {
        writeLEDCharcateristicValue(Data([0x0]))
    }
    
    private func writeLEDCharcateristicValue(_ value: Data) {
        guard let ledCharacteristic = ledCharacteristic else {
            print("LED characteristic is not present, nothing to be done")
            return
        }
        print("Writing LED value...")
        basePeripheral.writeValue(value, for: ledCharacteristic, type: .withResponse)
    }

    private func parseAdvertisementData(_ advertisementDictionary: [String : Any]) -> (String?, [CBUUID]?) {
        var advertisedName: String
        var advertisedServices: [CBUUID]

        if let name = advertisementDictionary[CBAdvertisementDataLocalNameKey] as? String{
            advertisedName = name
        } else {
            advertisedName = "N/A"
        }
        if let services = advertisementDictionary[CBAdvertisementDataServiceUUIDsKey] as? [CBUUID] {
            advertisedServices = services
        } else {
            advertisedServices = [CBUUID]()
        }
        
        return (advertisedName, advertisedServices)
    }
    
    // MARK: - NSObject protocols
    
    override func isEqual(_ object: Any?) -> Bool {
        if object is BlinkyPeripheral {
            let peripheralObject = object as! BlinkyPeripheral
            return peripheralObject.basePeripheral.identifier == basePeripheral.identifier
        } else if object is CBPeripheral {
            let peripheralObject = object as! CBPeripheral
            return peripheralObject.identifier == basePeripheral.identifier
        } else {
            return false
        }
    }
    
    // MARK: - CBPeripheralDelegate
    
    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        if characteristic == buttonCharacteristic {
            if let aValue = characteristic.value {
                didReceiveButtonNotificationWithValue(aValue)
            }
        } else if characteristic == ledCharacteristic {
            if let aValue = characteristic.value {
                didWriteValueToLED(aValue)
            }
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didUpdateNotificationStateFor characteristic: CBCharacteristic, error: Error?) {
        if characteristic == buttonCharacteristic {
            print("Button notifications enabled")
            readButtonValue()
            readLEDValue()
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        if let services = peripheral.services {
            for service in services {
                if service.uuid == BlinkyPeripheral.nordicBlinkyServiceUUID {
                    print("LED Button service found")
                    //Capture and discover all characteristics for the blinky service
                    blinkyService = service
                    discoverCharacteristicsForBlinkyService(blinkyService!)
                }
            }
        }
    }
    
    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        if service == blinkyService {
            if let characteristics = service.characteristics {
                for characteristic in characteristics {
                    if characteristic.uuid == BlinkyPeripheral.buttonCharacteristicUUID {
                        print("Button characteristic found")
                        buttonCharacteristic = characteristic
                        enableButtonNotifications(buttonCharacteristic!)
                    } else if characteristic.uuid == BlinkyPeripheral.ledCharacteristicUUID {
                        print("LED characteristic found")
                        ledCharacteristic = characteristic
                    }
                }
            }
        }
    }
    
    func peripheral(_ peripheral: CBPeripheral, didWriteValueFor characteristic: CBCharacteristic, error: Error?) {
        if characteristic == ledCharacteristic {
            print("Reading LED characteristic...")
            peripheral.readValue(for: ledCharacteristic!)
        }
    }
}
